---
title: "Identification of TF binding sites in Brassica Rapa based on PWM from Arabidopsis, using Biocondauctor/universalmotif"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
library(TFBSTools)
library(universalmotif)
library(rtracklayer)
library(Biostrings)
library(tidyverse)
```



Use "Brapa3_promoters_1500" for following analysis, to predict TF binding sites.

```{r eval=FALSE}
#output of Biostrings after promoter anotation
Brapa3_promoters_1500 <- readDNAStringSet("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters_1500.fa")

#create list of files
meme_files <- list.files("~/Work/2019/Analysis/TF_binding/motifs" ,pattern=".txt$", full.names = T, recursive=TRUE)
#upload into r
all_meme_data <- lapply(meme_files, function(x) read_meme(x))

all_results_scan <- data.frame()
for (i in 1:length(all_meme_data)) {
  
all_results_scan <- rbind(all_results_scan, scan_sequences(all_meme_data[i], Brapa3_promoters_1500, verbose = 1, progress = FALSE, threshold = 0.6, threshold.type = "logodds"))
}

head(all_results_scan)
tail(all_results_scan)

all_results_scan %>%  mutate(TF_family = motif) %>% mutate(TF = str_remove(motif, "\\..*$"))

all_results_scan_TF_family <- all_results_scan %>%  mutate(TF_family = str_remove(motif, "\\..*$"))

```


```{r eval=FALSE}
getwd()
dir.create("Brapa_3/analysed_data/")
write_csv(all_results_scan, "Brapa_3/analysed_data/TF_scan_results_Brapa3_promoters1500.csv" )
write_csv(all_results_scan_TF_family, "Brapa_3/analysed_data/TF_scan_results_Brapa3_promoters1500_TF_family.csv" )

```


* Processing motifs
 * Processing sequences
Non-standard letters found, these will be ignored * Scanning sequences for motifs
 * Processing results
 * Generating output
 
 
#Analysis of TF binding sites

##1. Calculate the total number of all binding sites per TF or TF_family on each promoter and their sum

```{r eval=FALSE}
Summary_scores_TF_per_promoter <- all_results_scan_TF_family %>%
  tbl_df() %>% 
  group_by(sequence, motif) %>%
  summarise(
    N = n(),
    sum_score_promoter = sum(score),
    min_score_promoter = min(score),
    max_score_promoter = max(score),
    TF_max_score = max(max.score)
  )
dim(Summary_scores_TF_per_promoter)
#[1] 4412681       7

Summary_scores_TF_family_per_promoter <- all_results_scan_TF_family %>%
  tbl_df() %>%
  group_by(sequence, TF_family) %>%
  summarise(
    N = n(),
    sum_score_promoter = sum(score),
    min_score_promoter = min(score),
    max_score_promoter = max(score),
    TF_max_score = max(max.score)
  )
dim(Summary_scores_TF_family_per_promoter)
#[1] 1085913       7
```


##2. Select the strongest binding site per promoter:
    1. Filter “score” == max.score, based on "score.pct" > 99% (score.pct = score/max.score)
        1. Calculate their num/promoter
        2. Calculate their sum/promoter

```{r eval=FALSE}
Summary_max_score_TF_per_promoter <- all_results_scan_TF_family %>%
  tbl_df() %>%
  filter(score.pct > 99) %>%
  group_by(sequence, motif) %>%
  summarise(
    N = n(),
    sum_TF_max_score_promoter = sum(score),#add mean!
    TF_max_score_adj = mean(score),
    TF_max_score = max(max.score)
  )
dim(Summary_max_score_TF_per_promoter)
#[1] [1] 304208      6

Summary_max_score_TF_family_per_promoter <- all_results_scan_TF_family %>%
  tbl_df() %>% 
  filter(score.pct > 99) %>%
  group_by(sequence, TF_family) %>%
  summarise(
    N = n(),
    sum_TF_max_score_promoter = sum(score),#add mean!
    TF_max_score_adj = mean(score),
    TF_max_score = max(max.score)
  )
dim(Summary_max_score_TF_family_per_promoter)
# [1] 257328      6
```

```{r eval=FALSE}
write_csv(Summary_scores_TF_per_promoter, "Brapa_3/analysed_data/Brapa3_Summary_scores_TF_per_promoter1500.csv" )
write_csv(Summary_max_score_TF_per_promoter, "Brapa_3/analysed_data/Brapa3_Summary_max_score_TF_per_promoter1500.csv" )
write_csv(Summary_scores_TF_family_per_promoter, "Brapa_3/analysed_data/Brapa3_Summary_scores_TF_family_per_promoter1500.csv" )
write_csv(Summary_max_score_TF_family_per_promoter, "Brapa_3/analysed_data/Brapa3_Summary_max_score_TF_family_per_promoter1500.csv" )
```





#####-----The End!-------########

######---------########### *Earlier version; if you start from the output of bedtools*

## Prepare DNAStringSet object from ~/Work/2019/Analysis/Promoters/BRapa/Brapa_promoters1500.fa for subsiquent analysis
Use tab.fa file to preprocess promoter DNAStringSet (sequences.fa file) for the following analysis
promoters1500_tab.fa and promoters1500.fa, contain the same info, different format

##modify promoters1500_tab.fa to prepare GenomiRanges object, in order to change the names of promoters1500.fa to be the gene name
change names(DNAStringSET) to include relevant gene names

```{r eval=FALSE}
Brapa3_promoters1500_txt <- tbl_df(read.delim("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters1500_tab.fa", header=FALSE, stringsAsFactors=FALSE))
Brapa3_promoters1500_annotation <- import.gff3("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters_1500.gff3")

Brapa3_promoters1500_bed <- readDNAStringSet("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters1500.fa")

#output of Biostrings
Brapa3_promoters_1500  <- readDNAStringSet("~/Work/2019/Analysis/Promoters/BRapa/Brapa_v3.0/promoters/Brapa_3.0_promoters_1500.fa")


```



```{r eval=FALSE}
# modify promoters1500.bed to prepare GenomiRanges object
Brapa3_promoters1500_txt <- Brapa3_promoters1500_txt %>%
  mutate(
    names = V1,
    V1 = str_replace_all(V1, ":", "_"),
    V1 = str_replace_all(V1, "\\(|\\)", ""),
    strand = str_extract(V1, "\\+$|\\-$"),
    V1 = str_replace_all(V1, "\\+$|\\-$", "")
  ) %>%
  separate(V1, into = c("seqnames", "start", "end")) %>% 
  mutate(start = as.numeric(start) + 1)

names(Brapa3_promoters1500_txt)[4] <- "sequence"
dim(Brapa3_promoters1500_txt)
#[1] 43963     6

Brapa3_promoters1500_txt_GR <-
  makeGRangesFromDataFrame(as.data.frame(Brapa3_promoters1500_txt),
                           keep.extra.columns = TRUE)


#------use overlap between promoters1500_txt_GR and promoters_1500.gff3 to match between them
#findOverlaps(query, subject)

#GenomicRanges::findOverlaps(Brapa_promoters1500_annotation, Brapa_promoters1500_txt_GR) 
Brapa3_promoters1500_txt_GR <- Brapa3_promoters1500_txt_GR[subjectHits(GenomicRanges::findOverlaps(Brapa3_promoters1500_annotation, Brapa3_promoters1500_txt_GR)),]

#----add ID column from promoters1500_annotation to promoters1500_txt_GR with gene name
mcols(Brapa3_promoters1500_txt_GR)$ID <-str_replace(mcols(Brapa3_promoters1500_annotation)$ID, "promoter:", "")

#Change names in DNAStringSet obgect to be gene names
Brapa3_promoters1500_bed <- Brapa3_promoters1500_bed[match(names(Brapa3_promoters1500_bed), mcols(Brapa3_promoters1500_txt_GR)$names),]
names(Brapa3_promoters1500_bed) <- mcols(Brapa3_promoters1500_txt_GR)$ID


# Test if output from bedtools and Biostrings are the same for negative strand:

#"BraA01g000020.3C"
Brapa3_promoters_1500[grep("BraA01g000020.3C", names(Brapa3_promoters_1500))]
mcols(Brapa3_promoters1500_txt_GR[strand(Brapa3_promoters1500_txt_GR) == "-",])[1,]

#"BraA01g020870.3C"
Brapa3_promoters_1500[grep("BraA01g020870.3C", names(Brapa3_promoters_1500))]
mcols(Brapa3_promoters1500_txt_GR[strand(Brapa3_promoters1500_txt_GR) == "-",])[1000,]

```
